File: app/src/main/java/de/hsesslingen/stundenplan/data/QisParser.kt
package de.hsesslingen.stundenplan.data

import org.jsoup.Jsoup
import org.jsoup.nodes.Element

/**
 * Parsing for HS Esslingen's QIS/LSF pages. The HTML is old-school table-layout markup with no
 * stable ids/classes beyond a handful of presentation classes ("plan1"/"plan2"/"notiz"/"klein"),
 * so parsing leans on structural selectors and regexes rather than ids.
 */
object QisParser {

    private val PARALLELID_RE = Regex("""k_parallel\.parallelid=(\d+)""")
    private val ABSTGVNR_RE = Regex("""k_abstgv\.abstgvnr=(\d+)""")
    private val TIME_RE = Regex("""(\d{1,2}:\d{2})\s*-\s*(\d{1,2}:\d{2})\s*(?:\(([^)]+)\))?""")

    /** Parses the public "Studiengangpläne (Liste)" catalog page into selectable entries. */
    fun parseStudiengangList(html: String): List<Studiengang> {
        val doc = Jsoup.parse(html)
        val header = doc.select("th").firstOrNull { it.text().contains("Studiengänge") } ?: return emptyList()
        val table = header.closest("table") ?: return emptyList()
        val result = mutableListOf<Studiengang>()
        for (row in table.select("tbody tr")) {
            val cells = row.select("> td")
            if (cells.size < 2) continue
            val code = cells[0].selectFirst("a")?.text()?.trim()?.takeIf { it.isNotBlank() } ?: continue
            val planHref = cells[1].selectFirst("a")?.attr("href") ?: continue
            val abstgvnr = ABSTGVNR_RE.find(planHref)?.groupValues?.get(1) ?: continue
            val parallelid = PARALLELID_RE.find(planHref)?.groupValues?.get(1) ?: continue
            result.add(Studiengang(code = code, abstgvnr = abstgvnr, parallelid = parallelid))
        }
        return result.distinctBy { it.id }.sortedBy { it.code }
    }

    /** Parses a "wplan" weekly timetable page (show=plan, P.vx=lang) into individual events. */
    fun parseTimetable(html: String): List<TimetableEvent> {
        val doc = Jsoup.parse(html)
        // The page nests tables for full-page layout, so anchor on the actual weekday header
        // cell and walk up to its immediate table rather than matching any ancestor table.
        val montagHeader = doc.select("th").firstOrNull { it.text().contains("Montag") }
            ?: error("Stundenplan-Tabelle wurde auf der Seite nicht gefunden (unerwartetes Seitenformat).")
        val table = montagHeader.closest("table")
            ?: error("Stundenplan-Tabelle wurde auf der Seite nicht gefunden (unerwartetes Seitenformat).")
        val grid = resolveGrid(table)

        val columnToDay = mapOf(
            2 to Weekday.MONDAY,
            3 to Weekday.TUESDAY,
            4 to Weekday.WEDNESDAY,
            5 to Weekday.THURSDAY,
            6 to Weekday.FRIDAY,
        )

        val events = mutableListOf<TimetableEvent>()
        for (cell in grid) {
            // Day cells are classed "plan2", or "plan22"/"plan222"/... for taller cells (QIS bumps
            // the digit count with the rowspan) — legend/axis cells ("plan1", "plan5-7") never
            // match this. Matching only the exact "plan2" class silently dropped every event sitting
            // in one of the taller cells (e.g. Tutorien scheduled in long afternoon blocks).
            if (!cell.element.className().matches(Regex("plan2+"))) continue
            val day = columnToDay[cell.col] ?: continue
            for (eventTable in cell.element.select("> table")) {
                parseEventTable(eventTable, day)?.let { events.add(it) }
            }
        }
        return events.distinct().sortedWith(compareBy({ it.day }, { it.startMinutes }))
    }

    private fun parseEventTable(eventTable: Element, day: Weekday): TimetableEvent? {
        val title = eventTable.selectFirst("td.klein a")?.text()?.clean()
            ?: eventTable.selectFirst("td.klein")?.text()?.clean()
            ?: return null

        var startMinutes: Int? = null
        var endMinutes: Int? = null
        var frequency: String? = null
        var startDate: String? = null
        var endDate: String? = null
        var room: String? = null
        var category: String? = null
        var lecturer: String? = null

        for (notiz in eventTable.select("td.notiz")) {
            val text = notiz.text().clean()
            if (text.isBlank()) continue

            if (startMinutes == null) {
                val m = TIME_RE.find(text)
                if (m != null) {
                    startMinutes = parseTimeToMinutes(m.groupValues[1])
                    endMinutes = parseTimeToMinutes(m.groupValues[2])
                    frequency = m.groupValues.getOrNull(3)?.clean()?.ifBlank { null }
                    continue
                }
            }
            if (text.startsWith("Start:", ignoreCase = true)) {
                startDate = text.substringAfter(":").clean().ifBlank { null }
                continue
            }
            if (text.startsWith("Ende:", ignoreCase = true)) {
                endDate = text.substringAfter(":").clean().ifBlank { null }
                continue
            }
            if (text.startsWith("Dozent", ignoreCase = true)) {
                lecturer = text.substringAfter(":").clean().ifBlank { null }
                continue
            }
            if (text.startsWith("Einrichtung", ignoreCase = true)) {
                continue
            }
            val roomLink = notiz.selectFirst("a")
            if (roomLink != null && room == null) {
                room = roomLink.text().clean().ifBlank { null }
                category = text.removePrefix(roomLink.text()).clean().trim(':', ' ').ifBlank { null }
                continue
            }
        }

        if (startMinutes == null || endMinutes == null) return null

        return TimetableEvent(
            day = day,
            title = title,
            startMinutes = startMinutes,
            endMinutes = endMinutes,
            frequency = frequency,
            room = room,
            lecturer = lecturer,
            category = category,
            startDate = startDate,
            endDate = endDate,
        )
    }

    private fun parseTimeToMinutes(hhmm: String): Int {
        val parts = hhmm.split(":")
        return parts[0].trim().toInt() * 60 + parts[1].trim().toInt()
    }

    private fun String.clean(): String = this.replace(' ', ' ').trim()

    // --- HTML table grid resolution (handles rowspan/colspan) ---

    private data class GridCell(val row: Int, val col: Int, val element: Element)

    private fun resolveGrid(table: Element): List<GridCell> {
        val rows = table.select("> tbody > tr, > tr").let { direct ->
            if (direct.isNotEmpty()) direct else table.select("tr")
        }
        val occupiedUntilRow = HashMap<Int, Int>()
        val result = mutableListOf<GridCell>()

        for ((r, tr) in rows.withIndex()) {
            var c = 0
            for (cell in tr.children()) {
                if (cell.tagName() != "td" && cell.tagName() != "th") continue
                while ((occupiedUntilRow[c] ?: -1) >= r) c++
                val colspan = cell.attr("colspan").toIntOrNull()?.coerceAtLeast(1) ?: 1
                val rowspan = cell.attr("rowspan").toIntOrNull()?.coerceAtLeast(1) ?: 1
                result.add(GridCell(r, c, cell))
                for (cc in c until c + colspan) {
                    occupiedUntilRow[cc] = r + rowspan - 1
                }
                c += colspan
            }
        }
        return result
    }
}

File: desktop/src/core/QisParser.cpp
#include "QisParser.h"

#include "HtmlDocument.h"

#include <QRegularExpression>
#include <QSet>
#include <algorithm>

namespace stundenplan::QisParser {

namespace {

const QRegularExpression kParallelIdRe(QStringLiteral("k_parallel\\.parallelid=(\\d+)"));
const QRegularExpression kAbstgvnrRe(QStringLiteral("k_abstgv\\.abstgvnr=(\\d+)"));
const QRegularExpression kTimeRe(
    QStringLiteral("(\\d{1,2}:\\d{2})\\s*-\\s*(\\d{1,2}:\\d{2})\\s*(?:\\(([^)]+)\\))?"));
const QRegularExpression kPlan2Re(QStringLiteral("^plan2+$"));

int parseTimeToMinutes(const QString &hhmm)
{
    const auto parts = hhmm.split(QLatin1Char(':'));
    return parts[0].trimmed().toInt() * 60 + parts[1].trimmed().toInt();
}

/** table.select("tbody tr") — rows nested (at any depth) inside a <tbody>, with a fallback to
 *  any descendant <tr> for pages whose HTML omits an explicit <tbody>. */
QList<HtmlElement> tbodyDescendantRows(const HtmlElement &table)
{
    QList<HtmlElement> result;
    for (const auto &tbody : table.selectDescendants(QStringLiteral("tbody")))
        result += tbody.selectDescendants(QStringLiteral("tr"));
    if (result.isEmpty())
        result = table.selectDescendants(QStringLiteral("tr"));
    return result;
}

/** table.select("> tbody > tr, > tr") falling back to table.select("tr") — used for grid
 *  resolution, which needs direct rows so nested tables (event blocks) aren't double-counted. */
QList<HtmlElement> directRowsOrFallback(const HtmlElement &table)
{
    QList<HtmlElement> direct;
    for (const auto &tbody : table.selectDirectChildren(QStringLiteral("tbody")))
        direct += tbody.selectDirectChildren(QStringLiteral("tr"));
    direct += table.selectDirectChildren(QStringLiteral("tr"));
    if (!direct.isEmpty())
        return direct;
    return table.selectDescendants(QStringLiteral("tr"));
}

struct GridCell {
    int row = 0;
    int col = 0;
    HtmlElement element;
};

QList<GridCell> resolveGrid(const HtmlElement &table)
{
    const auto rows = directRowsOrFallback(table);
    QHash<int, int> occupiedUntilRow;
    QList<GridCell> result;

    for (int r = 0; r < rows.size(); ++r) {
        int c = 0;
        for (const auto &cell : rows[r].children()) {
            const QString tag = cell.tagName();
            if (tag != QStringLiteral("td") && tag != QStringLiteral("th"))
                continue;
            while (occupiedUntilRow.value(c, -1) >= r)
                ++c;
            bool ok = false;
            int colspan = cell.attr(QStringLiteral("colspan")).toInt(&ok);
            if (!ok || colspan < 1)
                colspan = 1;
            int rowspan = cell.attr(QStringLiteral("rowspan")).toInt(&ok);
            if (!ok || rowspan < 1)
                rowspan = 1;
            result.append(GridCell{r, c, cell});
            for (int cc = c; cc < c + colspan; ++cc)
                occupiedUntilRow[cc] = r + rowspan - 1;
            c += colspan;
        }
    }
    return result;
}

std::optional<TimetableEvent> parseEventTable(const HtmlElement &eventTable, Weekday day)
{
    QString title;
    if (auto link = eventTable.firstDescendant(QStringLiteral("a")); link.isValid()) {
        // Only accept a link if it sits inside a td.klein — mirrors selectFirst("td.klein a").
        for (const auto &kleinTd : eventTable.selectDescendantsByClass(QStringLiteral("td"), QStringLiteral("klein"))) {
            if (auto a = kleinTd.firstDescendant(QStringLiteral("a")); a.isValid()) {
                title = a.text().trimmed();
                break;
            }
        }
    }
    if (title.isEmpty()) {
        const auto kleinTds = eventTable.selectDescendantsByClass(QStringLiteral("td"), QStringLiteral("klein"));
        if (!kleinTds.isEmpty())
            title = kleinTds.first().text().trimmed();
    }
    if (title.isEmpty())
        return std::nullopt;

    std::optional<int> startMinutes;
    std::optional<int> endMinutes;
    QString frequency, startDate, endDate, room, category, lecturer;

    for (const auto &notiz : eventTable.selectDescendantsByClass(QStringLiteral("td"), QStringLiteral("notiz"))) {
        const QString text = notiz.text().trimmed();
        if (text.isEmpty())
            continue;

        if (!startMinutes.has_value()) {
            auto match = kTimeRe.match(text);
            if (match.hasMatch()) {
                startMinutes = parseTimeToMinutes(match.captured(1));
                endMinutes = parseTimeToMinutes(match.captured(2));
                const QString freq = match.captured(3).trimmed();
                if (!freq.isEmpty())
                    frequency = freq;
                continue;
            }
        }
        if (text.startsWith(QStringLiteral("Start:"), Qt::CaseInsensitive)) {
            const QString value = text.mid(text.indexOf(QLatin1Char(':')) + 1).trimmed();
            if (!value.isEmpty())
                startDate = value;
            continue;
        }
        if (text.startsWith(QStringLiteral("Ende:"), Qt::CaseInsensitive)) {
            const QString value = text.mid(text.indexOf(QLatin1Char(':')) + 1).trimmed();
            if (!value.isEmpty())
                endDate = value;
            continue;
        }
        if (text.startsWith(QStringLiteral("Dozent"), Qt::CaseInsensitive)) {
            const QString value = text.mid(text.indexOf(QLatin1Char(':')) + 1).trimmed();
            if (!value.isEmpty())
                lecturer = value;
            continue;
        }
        if (text.startsWith(QStringLiteral("Einrichtung"), Qt::CaseInsensitive)) {
            continue;
        }
        auto roomLink = notiz.firstDescendant(QStringLiteral("a"));
        if (roomLink.isValid() && room.isEmpty()) {
            const QString linkText = roomLink.text().trimmed();
            room = linkText.isEmpty() ? QString() : linkText;
            QString rest = text;
            if (rest.startsWith(linkText))
                rest = rest.mid(linkText.length());
            rest = rest.trimmed();
            while (!rest.isEmpty() && (rest.front() == QLatin1Char(':') || rest.front() == QLatin1Char(' ')))
                rest.removeFirst();
            while (!rest.isEmpty() && (rest.back() == QLatin1Char(':') || rest.back() == QLatin1Char(' ')))
                rest.removeLast();
            if (!rest.isEmpty())
                category = rest;
            continue;
        }
    }

    if (!startMinutes.has_value() || !endMinutes.has_value())
        return std::nullopt;

    TimetableEvent event;
    event.day = day;
    event.title = title;
    event.startMinutes = *startMinutes;
    event.endMinutes = *endMinutes;
    event.frequency = frequency;
    event.room = room;
    event.lecturer = lecturer;
    event.category = category;
    event.startDate = startDate;
    event.endDate = endDate;
    return event;
}

} // namespace

QList<Studiengang> parseStudiengangList(const QString &html)
{
    HtmlDocument doc(html);
    if (!doc.isValid())
        return {};

    HtmlElement header;
    for (const auto &th : doc.root().selectDescendants(QStringLiteral("th"))) {
        if (th.text().contains(QStringLiteral("Studiengänge"))) {
            header = th;
            break;
        }
    }
    if (!header.isValid())
        return {};

    HtmlElement table = header.closest(QStringLiteral("table"));
    if (!table.isValid())
        return {};

    QList<Studiengang> result;
    QSet<QString> seenIds;
    for (const auto &row : tbodyDescendantRows(table)) {
        const auto cells = row.selectDirectChildren(QStringLiteral("td"));
        if (cells.size() < 2)
            continue;
        const auto codeLink = cells[0].firstDescendant(QStringLiteral("a"));
        const QString code = codeLink.isValid() ? codeLink.text().trimmed() : QString();
        if (code.isEmpty())
            continue;
        const auto planLink = cells[1].firstDescendant(QStringLiteral("a"));
        if (!planLink.isValid())
            continue;
        const QString href = planLink.attr(QStringLiteral("href"));
        const auto abstgvnrMatch = kAbstgvnrRe.match(href);
        const auto parallelIdMatch = kParallelIdRe.match(href);
        if (!abstgvnrMatch.hasMatch() || !parallelIdMatch.hasMatch())
            continue;

        Studiengang s;
        s.code = code;
        s.abstgvnr = abstgvnrMatch.captured(1);
        s.parallelid = parallelIdMatch.captured(1);
        if (seenIds.contains(s.id()))
            continue;
        seenIds.insert(s.id());
        result.append(s);
    }

    std::sort(result.begin(), result.end(), [](const Studiengang &a, const Studiengang &b) {
        return a.code < b.code;
    });
    return result;
}

QList<TimetableEvent> parseTimetable(const QString &html, QString *errorMessage)
{
    static const QString kFormatError =
        QStringLiteral("Stundenplan-Tabelle wurde auf der Seite nicht gefunden (unerwartetes Seitenformat).");

    HtmlDocument doc(html);
    if (!doc.isValid()) {
        if (errorMessage)
            *errorMessage = kFormatError;
        return {};
    }

    HtmlElement montagHeader;
    for (const auto &th : doc.root().selectDescendants(QStringLiteral("th"))) {
        if (th.text().contains(QStringLiteral("Montag"))) {
            montagHeader = th;
            break;
        }
    }
    if (!montagHeader.isValid()) {
        if (errorMessage)
            *errorMessage = kFormatError;
        return {};
    }
    HtmlElement table = montagHeader.closest(QStringLiteral("table"));
    if (!table.isValid()) {
        if (errorMessage)
            *errorMessage = kFormatError;
        return {};
    }

    const auto grid = resolveGrid(table);

    QHash<int, Weekday> columnToDay{
        {2, Weekday::Monday},
        {3, Weekday::Tuesday},
        {4, Weekday::Wednesday},
        {5, Weekday::Thursday},
        {6, Weekday::Friday},
    };

    QList<TimetableEvent> events;
    for (const auto &cell : grid) {
        // Day cells are classed "plan2", or "plan22"/"plan222"/... for taller (rowspan-bumped)
        // cells; legend/axis cells never match this.
        if (!kPlan2Re.match(cell.element.className()).hasMatch())
            continue;
        if (!columnToDay.contains(cell.col))
            continue;
        const Weekday day = columnToDay.value(cell.col);
        for (const auto &eventTable : cell.element.selectDirectChildren(QStringLiteral("table"))) {
            if (auto event = parseEventTable(eventTable, day))
                events.append(*event);
        }
    }

    std::sort(events.begin(), events.end(), [](const TimetableEvent &a, const TimetableEvent &b) {
        if (a.day != b.day)
            return static_cast<int>(a.day) < static_cast<int>(b.day);
        return a.startMinutes < b.startMinutes;
    });

    // distinct() — drop exact duplicates while preserving order.
    QList<TimetableEvent> distinct;
    for (const auto &event : events) {
        if (!distinct.contains(event))
            distinct.append(event);
    }
    return distinct;
}

} // namespace stundenplan::QisParser
